{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current kernel CWD: /beegfs/home/mlopezdecas/single_cell/conformalized_single_cell_annotator\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/beegfs/home/mlopezdecas/.conda/envs/mamba-env/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "\n",
    "import sys\n",
    "import os\n",
    "\n",
    "cwd = os.getcwd()\n",
    "print(\"Current kernel CWD:\", cwd)\n",
    "\n",
    "## 2a. Opción: ir 'hacia abajo' al subdirectorio donde está tu módulo\n",
    "#module_dir = os.path.join(cwd, \"conformalized_single_cell_annotator\")\n",
    "#sys.path.insert(0, module_dir)\n",
    "\n",
    "\n",
    "import numpy as np  \n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "from conformalSC_annotator import  ConformalSCAnnotator\n",
    "from torchcp.classification.score import  APS,RAPS, THR\n",
    "\n",
    "\n",
    "# First we read the already preprocessed query data using scanpy. In this case data is coming from a h5ad file.\n",
    "\n",
    "query_data_path = '../test_data/gastrulation/gastrulation_query_1.h5ad'\n",
    "\n",
    "adata_query = sc.read_h5ad(query_data_path) \n",
    "\n",
    "## We need a .var column that contains the gene names (if not created).\n",
    "## Sometimes this information is on index column adata_query.var_names, but we explicity in a new column if not exist .\n",
    "## In this case, we suppose that the column is already created and named: \"features\".\n",
    "\n",
    "gene_names_column = \"features\" \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(self, X, var_query=None, obs_query=None, var_query_gene_column_name='gene_name', underlying_model='torch_net')\n"
     ]
    }
   ],
   "source": [
    "import inspect\n",
    "print(inspect.signature(ConformalSCAnnotator.__init__))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data stored in adata.X will be used for annotation\n",
      "Model not trained yet. Fitting the model first.\n",
      "Loading reference data...\n",
      "Reference data shape: (1843, 2000)\n",
      "\n",
      "Reference data loaded.\n",
      "\n",
      "Initial reference data label distribution:\n",
      "celltype\n",
      "Rostral neurectoderm              300\n",
      "Epiblast                          283\n",
      "Mesenchyme                        218\n",
      "Mixed mesoderm                    190\n",
      "Nascent mesoderm                  160\n",
      "Primitive Streak                  143\n",
      "Visceral endoderm                 104\n",
      "Caudal epiblast                    80\n",
      "Def. endoderm                      79\n",
      "Haematoendothelial progenitors     69\n",
      "Blood progenitors 2                56\n",
      "Gut                                56\n",
      "Surface ectoderm                   53\n",
      "Blood progenitors 1                52\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Query data label distribution:\n",
      "celltype\n",
      "Rostral neurectoderm              736\n",
      "Epiblast                          641\n",
      "Nascent mesoderm                  485\n",
      "Primitive Streak                  395\n",
      "Mixed mesoderm                    390\n",
      "Mesenchyme                        379\n",
      "Caudal epiblast                   338\n",
      "Blood progenitors 2               212\n",
      "Haematoendothelial progenitors    185\n",
      "Def. endoderm                     156\n",
      "Blood progenitors 1               149\n",
      "Visceral endoderm                 128\n",
      "Gut                               108\n",
      "Surface ectoderm                   92\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Excluding cell types in: 45:\n",
      "[]\n",
      "Detecting common genes...\n",
      "Common genes detected: 2000\n",
      "Train data shape: (1056, 2000)\n",
      "Validation data shape: (249, 2000)\n",
      "Test data shape: (185, 2000)\n",
      "Calibration data shape: (353, 2000)\n",
      "\n",
      "\n",
      "Data loaded!\n",
      "\n",
      "Training OOD detector...\n",
      "Epoch 1, Training Loss: 0.4688, Validation Loss: 0.4132\n",
      "Epoch 2, Training Loss: 0.4310, Validation Loss: 0.3986\n",
      "Finished training: train_loss=0.4310, val_loss=0.3986\n",
      "Fitted the one-class model\n",
      "Calibrated! 353\n",
      "OOD detector trained!\n",
      "Training scmap model...\n",
      "scmap is ready to predict.\n",
      "Calibrating the model...\n",
      "Using standard taxonomy\n",
      "Using standard taxonomy\n",
      "Using standard taxonomy\n",
      "Model calibrated.\n",
      "\n",
      "Test Accuracy: 76.22%\n",
      "Precision (Weighted Average): 76.32%\n",
      "Recall (Weighted Average): 76.22%\n",
      "F1 Score (Weighted Average): 75.36%\n",
      "\n",
      "Scores by Class:\n",
      "  Class 0: Precision: 0.67, Recall: 0.40, F1 Score: 0.50\n",
      "  Class 1: Precision: 0.83, Recall: 0.83, F1 Score: 0.83\n",
      "  Class 2: Precision: 0.57, Recall: 0.50, F1 Score: 0.53\n",
      "  Class 3: Precision: 0.70, Recall: 0.88, F1 Score: 0.78\n",
      "  Class 4: Precision: 0.67, Recall: 0.93, F1 Score: 0.78\n",
      "  Class 5: Precision: 0.67, Recall: 0.33, F1 Score: 0.44\n",
      "  Class 6: Precision: 0.78, Recall: 1.00, F1 Score: 0.88\n",
      "  Class 7: Precision: 1.00, Recall: 1.00, F1 Score: 1.00\n",
      "  Class 8: Precision: 0.83, Recall: 0.79, F1 Score: 0.81\n",
      "  Class 9: Precision: 0.69, Recall: 0.69, F1 Score: 0.69\n",
      "  Class 10: Precision: 0.58, Recall: 0.50, F1 Score: 0.54\n",
      "  Class 11: Precision: 0.75, Recall: 0.60, F1 Score: 0.67\n",
      "  Class 12: Precision: 1.00, Recall: 0.80, F1 Score: 0.89\n",
      "  Class 13: Precision: 0.92, Recall: 1.00, F1 Score: 0.96\n",
      "\n",
      "Performing conformal prediction...\n",
      "\n",
      "Conformal predictor 0.01 - Coverage Rate: 1.0, CovGap: 0.9999990463256836, Average Size: 14.0\n",
      "Size Distribution: {14: 185}\n",
      "\n",
      "Conformal predictor 0.05 - Coverage Rate: 1.0, CovGap: 5.000000953674316, Average Size: 14.0\n",
      "Size Distribution: {14: 185}\n",
      "\n",
      "Conformal predictor 0.1 - Coverage Rate: 0.8864864706993103, CovGap: 10.949247360229492, Average Size: 1.4540539979934692\n",
      "Size Distribution: {0: 1, 1: 105, 2: 73, 3: 6}\n",
      "Starting annotation process...\n",
      "\n",
      "Performing OOD detection...\n",
      "OOD samples detected: 304\n",
      "\n",
      "Performing conformal prediction...\n",
      "\n",
      "Performing conformal prediction: Done\n",
      "\n",
      "Annotation process completed.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "model = \"scmap\"  # Choose the underlying model for the annotator. Options: \"torch_net\", \"celltypist\", \"scmap\". Default: \"torch_net\"\n",
    "\n",
    "annotator = ConformalSCAnnotator(adata_query,\n",
    "                                var_query_gene_column_name = gene_names_column, \n",
    "                                underlying_model = model) # obs_query is optional, it will be used for annotate the predicted cells. \n",
    "\n",
    "\n",
    "\n",
    "# Define que network architecture if torch_net is used as underlying model.   \n",
    "network_architecture:dict = {   \n",
    "            \"hidden_sizes\": [128, 128, 64, 64],\n",
    "            \"dropout_rates\": [0.4, 0.3, 0.4, 0.25],\n",
    "            \"learning_rate\": 0.0001}\n",
    "\n",
    "\n",
    "OOD_detector_config = { \"pvalues\": \"marginal\",             # choose between marginal or conditional. Def: \"marginal\"\n",
    "                        \"alpha\": 0.1,                      # Significance level for the hyoothesis test\n",
    "                        \"delta\": 0.1,                      # only for conditional pvalues\n",
    "                        \"hidden_sizes\": [ 556,  124, 64],      # AE hidden sizes and topology of the network\n",
    "                        \"dropout_rates\": [ 0.2,  0.25, 0.25],\n",
    "                        \"learning_rate\": 0.0001,\n",
    "                        \"batch_size\": 32,\n",
    "                        \"n_epochs\": 2}\n",
    "    \n",
    "\n",
    "reference_data_path = '../test_data/gastrulation/gastrulation_reference_1.h5ad'     # Path to the reference data\n",
    "\n",
    "\n",
    "annotator.configure(reference_path = reference_data_path,        # Path to the reference data in format .h5ad or adata\n",
    "                    model_architecture = network_architecture,   # Optional, if not provided, default values will be used\n",
    "                    OOD_detector = OOD_detector_config,          # Optional, if not provided, default values will be used\n",
    "                    CP_predictor = \"standard\",                   # standard, mondrian or cluster\n",
    "                    cell_names_column = \"celltype\",               # class name for fitting the model.  cell_type or celltype_level3, celltype \n",
    "                    cell_types_excluded_treshold = 45,           # Exclude cell types with less than 50 cells\n",
    "                    test =  True,                                # Perform internal test of the model\n",
    "                    alpha = [0.01, 0.05, 0.1],                   # Confidence of the predictions (can be a single element)\n",
    "                    non_conformity_function = APS(),             # NC-function provided by or compatible with torchCP   (APS, RAPS, THR) \n",
    "                    epoch=200,\n",
    "                    batch_size = 42,\n",
    "                    random_state = None)  \n",
    "    \n",
    "\n",
    "# Annotate the query data. \n",
    "# If batch corrected data is available, it can be used .\n",
    "\n",
    "annotator.annotate(obsm_layer=None)  # If in .obsm some reduction of integration is conducted: None, \"X_pca_harmony\" or \"X_pca\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Predicted annotations sets: \n",
      "                    barcode  sample stage sequencing.batch theiler  \\\n",
      "index                                                               \n",
      "cell_43479  AAACATACCATGCA      19  E7.5                2    TS11   \n",
      "cell_43481  AAACATACGGTACT      19  E7.5                2    TS11   \n",
      "cell_43483  AAACATACTTGACG      19  E7.5                2    TS11   \n",
      "cell_43484  AAACATTGACGGAG      19  E7.5                2    TS11   \n",
      "cell_43487  AAACATTGGGTGAG      19  E7.5                2    TS11   \n",
      "...                    ...     ...   ...              ...     ...   \n",
      "cell_52454  TTTAGGCTGTTCAG      20  E7.5                2    TS11   \n",
      "cell_52457  TTTATCCTCTGTTT      20  E7.5                2    TS11   \n",
      "cell_52458  TTTCAGTGACTACG      20  E7.5                2    TS11   \n",
      "cell_52464  TTTGCATGACGCTA      20  E7.5                2    TS11   \n",
      "cell_52465  TTTGCATGGACGTT      20  E7.5                2    TS11   \n",
      "\n",
      "            doub.density  doublet  cluster  cluster.sub  cluster.stage  ...  \\\n",
      "index                                                                   ...   \n",
      "cell_43479      0.132363    False     17.0          2.0            7.0  ...   \n",
      "cell_43481      0.281227    False     10.0          1.0            9.0  ...   \n",
      "cell_43483      0.000000    False     11.0          3.0           11.0  ...   \n",
      "cell_43484      0.509998    False     17.0          1.0            7.0  ...   \n",
      "cell_43487      0.401380    False     19.0          1.0            5.0  ...   \n",
      "...                  ...      ...      ...          ...            ...  ...   \n",
      "cell_52454      0.697361    False     10.0          3.0            9.0  ...   \n",
      "cell_52457      0.432823    False      1.0          3.0            3.0  ...   \n",
      "cell_52458      0.476887    False     15.0          1.0            9.0  ...   \n",
      "cell_52464      0.480761    False      2.0          6.0           13.0  ...   \n",
      "cell_52465      0.492117    False      2.0          3.0            9.0  ...   \n",
      "\n",
      "            log1p_total_counts_ribo  pct_counts_ribo total_counts_hb  \\\n",
      "index                                                                  \n",
      "cell_43479                      0.0              0.0             0.0   \n",
      "cell_43481                      0.0              0.0             0.0   \n",
      "cell_43483                      0.0              0.0             0.0   \n",
      "cell_43484                      0.0              0.0             0.0   \n",
      "cell_43487                      0.0              0.0             0.0   \n",
      "...                             ...              ...             ...   \n",
      "cell_52454                      0.0              0.0             0.0   \n",
      "cell_52457                      0.0              0.0             0.0   \n",
      "cell_52458                      0.0              0.0             0.0   \n",
      "cell_52464                      0.0              0.0             0.0   \n",
      "cell_52465                      0.0              0.0             0.0   \n",
      "\n",
      "           log1p_total_counts_hb  pct_counts_hb  \\\n",
      "index                                             \n",
      "cell_43479                   0.0            0.0   \n",
      "cell_43481                   0.0            0.0   \n",
      "cell_43483                   0.0            0.0   \n",
      "cell_43484                   0.0            0.0   \n",
      "cell_43487                   0.0            0.0   \n",
      "...                          ...            ...   \n",
      "cell_52454                   0.0            0.0   \n",
      "cell_52457                   0.0            0.0   \n",
      "cell_52458                   0.0            0.0   \n",
      "cell_52464                   0.0            0.0   \n",
      "cell_52465                   0.0            0.0   \n",
      "\n",
      "                          predicted_labels  \\\n",
      "index                                        \n",
      "cell_43479                      Mesenchyme   \n",
      "cell_43481            Rostral neurectoderm   \n",
      "cell_43483                             OOD   \n",
      "cell_43484                      Mesenchyme   \n",
      "cell_43487  Haematoendothelial progenitors   \n",
      "...                                    ...   \n",
      "cell_52454            Rostral neurectoderm   \n",
      "cell_52457            Rostral neurectoderm   \n",
      "cell_52458                Surface ectoderm   \n",
      "cell_52464                        Epiblast   \n",
      "cell_52465                        Epiblast   \n",
      "\n",
      "                                         prediction_sets_0.01  \\\n",
      "index                                                           \n",
      "cell_43479  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_43481  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_43483                                              [OOD]   \n",
      "cell_43484  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_43487  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "...                                                       ...   \n",
      "cell_52454  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_52457  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_52458  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_52464  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_52465  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "\n",
      "                                         prediction_sets_0.05  \\\n",
      "index                                                           \n",
      "cell_43479  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_43481  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_43483                                              [OOD]   \n",
      "cell_43484  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_43487  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "...                                                       ...   \n",
      "cell_52454  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_52457  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_52458  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_52464  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "cell_52465  [Blood progenitors 1, Blood progenitors 2, Cau...   \n",
      "\n",
      "                                     prediction_sets_0.1  \\\n",
      "index                                                      \n",
      "cell_43479                                  [Mesenchyme]   \n",
      "cell_43481                        [Rostral neurectoderm]   \n",
      "cell_43483                                         [OOD]   \n",
      "cell_43484                                  [Mesenchyme]   \n",
      "cell_43487  [Haematoendothelial progenitors, Mesenchyme]   \n",
      "...                                                  ...   \n",
      "cell_52454       [Caudal epiblast, Rostral neurectoderm]   \n",
      "cell_52457                        [Rostral neurectoderm]   \n",
      "cell_52458                            [Surface ectoderm]   \n",
      "cell_52464                                    [Epiblast]   \n",
      "cell_52465                                    [Epiblast]   \n",
      "\n",
      "                                           predicted_NCscores  \n",
      "index                                                          \n",
      "cell_43479  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.12352579...  \n",
      "cell_43481  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...  \n",
      "cell_43483  [nan, nan, nan, nan, nan, nan, nan, nan, nan, ...  \n",
      "cell_43484  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.44802629...  \n",
      "cell_43487  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.6873674988746...  \n",
      "...                                                       ...  \n",
      "cell_52454  [1.0, 1.0, 0.932097315788269, 1.0, 1.0, 1.0, 1...  \n",
      "cell_52457  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...  \n",
      "cell_52458  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, ...  \n",
      "cell_52464  [1.0, 1.0, 1.0, 1.0, 0.7297784686088562, 1.0, ...  \n",
      "cell_52465  [1.0, 1.0, 1.0, 1.0, 0.978469729423523, 1.0, 1...  \n",
      "\n",
      "[4394 rows x 50 columns]\n"
     ]
    }
   ],
   "source": [
    "# The annotations are provided in the original query data object:\n",
    "\n",
    "annotated_cells = annotator.adata_query.obs\n",
    "print(\"\\nPredicted annotations sets: \\n\" , annotated_cells)\n",
    "\n",
    "# And the results of the internal test:\n",
    "\n",
    "test_results = annotator.test_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# We can get the results from the adata object and store in a classical df:\n",
    "# predicted labels sntands for the predictions of the underlying model without conformal prediction.\n",
    "\n",
    "results = []\n",
    "for pred,cp_pred_001,cp_pred_005, cp_pred_010 in zip(\n",
    "        annotator.adata_query.obs[\"predicted_labels\"],\n",
    "        annotator.adata_query.obs[\"prediction_sets_0.01\"],\n",
    "        annotator.adata_query.obs[\"prediction_sets_0.05\"],\n",
    "        annotator.adata_query.obs[\"prediction_sets_0.1\"] ):\n",
    "        \n",
    "        \n",
    "    #print(f\"Predicted: {pred} - CP 0.01: {cp_pred_001} - CP 0.05: {cp_pred_005} - CP 0.10: {cp_pred_010}\")\n",
    "        \n",
    "    results.append({\n",
    "        \"Predicted\": pred,\n",
    "        \"CP 0.01\": cp_pred_001,\n",
    "        \"CP 0.05\": cp_pred_005,\n",
    "        \"CP 0.1\": cp_pred_010\n",
    "    })\n",
    "    \n",
    "\n",
    "df_results = pd.DataFrame(results)\n",
    "#df_results.to_csv(\"saved_Results.csv\", index=False)  # Save to CSV if needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Predicted</th>\n",
       "      <th>CP 0.01</th>\n",
       "      <th>CP 0.05</th>\n",
       "      <th>CP 0.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mesenchyme</td>\n",
       "      <td>[Mesenchyme]</td>\n",
       "      <td>[Mesenchyme]</td>\n",
       "      <td>[Mesenchyme]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Rostral neurectoderm</td>\n",
       "      <td>[Rostral neurectoderm]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Rostral neurectoderm]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OOD</td>\n",
       "      <td>[OOD]</td>\n",
       "      <td>[OOD]</td>\n",
       "      <td>[OOD]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OOD</td>\n",
       "      <td>[OOD]</td>\n",
       "      <td>[OOD]</td>\n",
       "      <td>[OOD]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Haematoendothelial progenitors</td>\n",
       "      <td>[Haematoendothelial progenitors, Surface ectod...</td>\n",
       "      <td>[Haematoendothelial progenitors]</td>\n",
       "      <td>[Haematoendothelial progenitors]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OOD</td>\n",
       "      <td>[OOD]</td>\n",
       "      <td>[OOD]</td>\n",
       "      <td>[OOD]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Mesenchyme</td>\n",
       "      <td>[Blood progenitors 1, Gut, Haematoendothelial ...</td>\n",
       "      <td>[Haematoendothelial progenitors, Mesenchyme]</td>\n",
       "      <td>[Haematoendothelial progenitors, Mesenchyme]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Rostral neurectoderm</td>\n",
       "      <td>[Rostral neurectoderm, Surface ectoderm]</td>\n",
       "      <td>[Rostral neurectoderm]</td>\n",
       "      <td>[Rostral neurectoderm]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Haematoendothelial progenitors</td>\n",
       "      <td>[Blood progenitors 1, Haematoendothelial proge...</td>\n",
       "      <td>[Haematoendothelial progenitors]</td>\n",
       "      <td>[Haematoendothelial progenitors]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Nascent mesoderm</td>\n",
       "      <td>[Mixed mesoderm, Nascent mesoderm, Primitive S...</td>\n",
       "      <td>[Mixed mesoderm, Nascent mesoderm]</td>\n",
       "      <td>[Nascent mesoderm]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Predicted  \\\n",
       "0                      Mesenchyme   \n",
       "1            Rostral neurectoderm   \n",
       "2                             OOD   \n",
       "3                             OOD   \n",
       "4  Haematoendothelial progenitors   \n",
       "5                             OOD   \n",
       "6                      Mesenchyme   \n",
       "7            Rostral neurectoderm   \n",
       "8  Haematoendothelial progenitors   \n",
       "9                Nascent mesoderm   \n",
       "\n",
       "                                             CP 0.01  \\\n",
       "0                                       [Mesenchyme]   \n",
       "1                             [Rostral neurectoderm]   \n",
       "2                                              [OOD]   \n",
       "3                                              [OOD]   \n",
       "4  [Haematoendothelial progenitors, Surface ectod...   \n",
       "5                                              [OOD]   \n",
       "6  [Blood progenitors 1, Gut, Haematoendothelial ...   \n",
       "7           [Rostral neurectoderm, Surface ectoderm]   \n",
       "8  [Blood progenitors 1, Haematoendothelial proge...   \n",
       "9  [Mixed mesoderm, Nascent mesoderm, Primitive S...   \n",
       "\n",
       "                                        CP 0.05  \\\n",
       "0                                  [Mesenchyme]   \n",
       "1                                            []   \n",
       "2                                         [OOD]   \n",
       "3                                         [OOD]   \n",
       "4              [Haematoendothelial progenitors]   \n",
       "5                                         [OOD]   \n",
       "6  [Haematoendothelial progenitors, Mesenchyme]   \n",
       "7                        [Rostral neurectoderm]   \n",
       "8              [Haematoendothelial progenitors]   \n",
       "9            [Mixed mesoderm, Nascent mesoderm]   \n",
       "\n",
       "                                         CP 0.1  \n",
       "0                                  [Mesenchyme]  \n",
       "1                        [Rostral neurectoderm]  \n",
       "2                                         [OOD]  \n",
       "3                                         [OOD]  \n",
       "4              [Haematoendothelial progenitors]  \n",
       "5                                         [OOD]  \n",
       "6  [Haematoendothelial progenitors, Mesenchyme]  \n",
       "7                        [Rostral neurectoderm]  \n",
       "8              [Haematoendothelial progenitors]  \n",
       "9                            [Nascent mesoderm]  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_results.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0.9324392080307007,\n",
       " 0.994827151298523,\n",
       " 0.9989463686943054,\n",
       " 0.996710479259491,\n",
       " 0.9997898936271667,\n",
       " 0.9871706366539001,\n",
       " 0.5648959875106812,\n",
       " 0.974643886089325,\n",
       " 0.9802080392837524,\n",
       " 0.9994446039199829,\n",
       " 0.9996454119682312,\n",
       " 0.9999656081199646,\n",
       " 0.9975572824478149,\n",
       " 0.9892067909240723]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
